python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py

```diff
@@ -0,0 +1,173 @@
+import sys
+from abc import abstractmethod
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Mapping,
+    TypeVar,
+    Union,
+    get_args,
+    get_origin,
+)
+
+if sys.version_info >= (3, 9):
+    from typing import TypeAlias
+else:
+    from typing_extensions import TypeAlias
+
+import polars as pl
+from dagster import InputContext, OutputContext
+
+if TYPE_CHECKING:
+    from upath import UPath
+
+
+T = TypeVar("T")
+
+
+# dump_to_path signature
+F_D: TypeAlias = Callable[[OutputContext, T, "UPath"], None]
+
+# load_from_path signature
+F_L: TypeAlias = Callable[["UPath", InputContext], T]
+
+
+class BaseTypeRouter(Generic[T]):
+    """Specifies how to apply a given dump/load operation to a given type annotation.
+    This base class trivially calls the dump/load functions if the type matches the most simple cases.
+    """
+
+    def __init__(self, context: Union[InputContext, OutputContext], typing_type: Any):
+        self.context = context
+        self.typing_type = typing_type
+
+    @abstractmethod
+    def match(self) -> bool:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def is_root_type(self) -> bool:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def parent_type(self) -> Any:
+        raise NotImplementedError
+
+    @property
+    def parent_type_router(self) -> "TypeRouter":
+        return resolve_type_router(self.context, self.parent_type)
+
+    @property
+    def is_eager(self) -> bool:
```

Comment:
I think this feels a bit over-specified to the polars usecase. I think this sort of logic can just live within the IOManager (i.e. the io manger just accesses `type_router.root_type` and compares that to a list of types).

Also a minor quibble, but I think this terminology of "root" and "parent" is a bit confusing (to me), and almost feels reversed. i.e. if you have a type `Optional[Something[int]]`, I personally think of `Optional` as the "root type", and `Something` as the "parent" of `int`, which is the reverse of what's going on in this PR.

Any objections to "`parent_type` -> `inner_type`" and `root_type` -> `base_type`?

Reply:
Moved to `BasePolarsUPathIOManager.type_router_is_eager`